File: Algorithms/Ticks/Tick1/maxheap.py

def heapify(x):
    for k in range(len(x)//2, -1, -1):
        maxheapify(x, k)

def maxheapify(x, ind): 
    
    l = 2*ind + 1
    r = 2*ind + 2
    largest = ind
        
    if l < len(x) and x[l] > x[ind]:
        largest = l

    if r < len(x) and x[r] > x[largest]:
        largest = r

    if largest != ind:
        x[largest], x[ind] = x[ind], x[largest]
        maxheapify(x, largest)     
        
def push (x, e):
    return heapify(x + [e])

def popmax(x):
    if len(x) is 0:
        raise IndexError()
    else:
        head = x.pop(0)
        heapify(x)
        return head

def main():
    test = [1,12,23,56,2,3,2,7,9,8,6,4,2,31,5,3,4]
    print("Pre-heapify: " + ' '.join(str(e) for e in test))
    heapify(test)
    print("Post-heapify: " + ' '.join(str(e) for e in test))
    push(test, 5)
    print("Push 5: " + ' '.join(str(e) for e in test))
    popmax(test)
    print("Remove max: " + ' '.join(str(e) for e in test))

if __name__ == "__main__":
    main()

File: Algorithms/Ticks/Tick1Star/MaxHeap.java

package Tick1Star;

import java.util.Arrays;
import java.util.List;
import java.util.Collections;
import java.util.ArrayList;
import java.lang.Comparable;

public class MaxHeap<T extends Comparable<T>> implements MaxHeapInterface<T> {

	private ArrayList<T> heap;

	public MaxHeap(List<T> x) {
		
		heap = new ArrayList<T>();
		for (T obj : x) {
			heap.add(obj);
		}

		heapify();
	}

	public void heapify() {

		int n = getLength();

		for (int k = n/2; k >= 0; k--) {
			maxHeapify(k);
		}

	}

	public void maxHeapify(int ind) {

		int n = getLength();
		int l = 2*ind + 1;
		int r = 2*ind + 2;
		int largest = ind;

		if (l < n && heap.get(l).compareTo(heap.get(largest)) == 1)  largest = l;
		if (r < n && heap.get(r).compareTo(heap.get(largest)) == 1) largest = r;

		if (largest != ind) {
			swap(largest, ind);
			maxHeapify(largest);
		}

	}

	public void swap(int j, int k) {
		
		T temp = heap.get(j);
		heap.set(j, heap.get(k));
		heap.set(k, temp);

	}

	public void insert(T e) {
		
		heap.add(e);	
		heapify();
	}

	public T popMax() throws EmptyHeapException {

		if (getLength() < 1) {
			throw new EmptyHeapException();
		}

		T head = heap.get(0);
		swap(0, getLength() - 1);
		heap.remove(getLength() - 1);	
		heapify();
		return head;

	}

	public int getLength() {
		return heap.size();
	}

	public void print() {
		
		for (int i = 0; i < getLength(); i++) {
			System.out.print(heap.get(i).toString() + " ");
		}

		System.out.println();
	}


}

